app.go

```diff
@@ -680,12 +680,18 @@ func (a *Application) processTx(txBytes []byte, isCheckTx bool) (TxHandlerResult
 
 	r, err := a.TxHandler.ProcessTx(state, txBytes, isCheckTx)
 	if err != nil {
+		// clear event cache on failed txs
+		a.EventHandler.Purge()
```

Comment:
Should use the same commit/rollback pattern as the receipt handler.
Call `defer a.EventHandler.Rollback()` before calling `a.TxHandler.ProcessTx`.

Reply:
Cool, fixed.